File: bfs/bfs.py
# Tree Structure:
#
#                     A
#                  /     \
#                B         C
#              /   \     /   \
#            D      E   F     G
#           /      / \   \
#          H      J   K   L
#

tree_test = {
    'A': ['B', 'C'],
    'B': ['D', 'E'],
    'C': ['F', 'G'],
    'D': ['H'],
    'E': ['J', 'K'],
    'F': ['L'],
    'G': [],
    'H': [], 'J': [],
    'K': [], 'L': [],
}

# funkcja do załadowania grafu
def load_graph():
    graph = {}
    # otworzenie pliku
    with open("D:/CODE/OESK/graph_with_weights.txt", "r") as file:
        for line in file:
            # jesli zaczynamy linie od # lub są puste linie to skip
            if line.startswith("#") or not line.strip():
                continue
            # dzielimy każdy node na dwie różne listy
            parts = line.split()
            # jeśli byłby błąd i pojawiłyby się więcej lub mniej niz 3 to skip
            if len(parts) != 3:
                continue
            #bierzemy tutaj 3 rzeczy ale uzywamy tylko 2
            from_node, to_node, weight = parts
            # zmieniamy na inty
            from_node = int(from_node)
            to_node = int(to_node)

            # jesli klucza nie ma w grafie to dodajemy
            if from_node not in graph:
                graph[from_node] = []
            # kazda wartosc do klucza dodajemy
            graph[from_node].append(to_node)
            # jeśli dziecko jest ostatnie to tworzymy je rodzicem z pustym zbiorem
            if to_node not in graph:
                graph[to_node] = []

    return graph

from collections import deque

def bfs(start_node):
    graph = load_graph()

    visited = set()
    visited.add(start_node)
    
    queue = deque([start_node])
    
    # 2. Store order of visited nodes
    traversal_order = []

    while queue:
        current_node = queue.popleft()
        traversal_order.append(current_node)

        # 3. Direct access & mark-on-push
        for neighbor in graph[current_node]:
            if neighbor not in visited:
                visited.add(neighbor)
                queue.append(neighbor)
                
    return traversal_order

result = bfs(0)
print(f"Total nodes visited: {len(result)}")
print(f"Last 5 nodes: {result[-5:]}")
print("koniec")

File: bfs/bfs.js
const fs = require("fs");

function loadGraph() {
    const graph = {};
    const data = fs.readFileSync("D:/CODE/OESK/graph_with_weights.txt", "utf-8");
    const lines = data.split(/\r?\n/);

    for (const line of lines) {
        if (line.startsWith("#") || !line.trim()) continue;
        const parts = line.trim().split(/\s+/);
        if (parts.length !== 3) continue;
        let [fromNode, toNode, weight] = parts;
        fromNode = parseInt(fromNode, 10);
        toNode = parseInt(toNode, 10);
        if (!graph[fromNode]) graph[fromNode] = [];
        graph[fromNode].push(toNode);
        if (!graph[toNode]) graph[toNode] = [];
    }
    return graph;
}

function bfs(start) {
    graph = loadGraph();
    const visited = new Set();
    const queue = [start];
    visited.add(start);

    let head = 0;

    while (head < queue.length) {
        const node = queue[head++]; 
        const neighbors = graph[node];

        if (neighbors) {
            for (let i = 0; i < neighbors.length; i++) {
                const neighbor = neighbors[i];
                if (!visited.has(neighbor)) {
                    visited.add(neighbor);
                    queue.push(neighbor);
                }
            }
        }
    }
    
    return queue.slice(-5);
}

const lastNodes = bfs(0);

console.log("Last 5 nodes:", lastNodes);
console.log("koniec");